homeassistant/components/unifiprotect/migrate.py

```diff
@@ -81,3 +86,68 @@ async def async_migrate_buttons(
 
     if count < len(to_migrate):
         _LOGGER.warning("Failed to migate %s reboot buttons", len(to_migrate) - count)
+
+
+async def async_migrate_device_ids(
+    hass: HomeAssistant, entry: ConfigEntry, protect: ProtectApiClient
+) -> None:
+    """
+    Migrate unique IDs from {device_id}_{name} format to {mac}_{name} format.
+
+    This makes devices persist better with in HA. Anything a device is unadopted/readopted or
+    the Protect instance has to rebuild the disk array, the device IDs of Protect devices
+    can change. This causes a ton of orphaned entities and loss of historical data. MAC
+    addresses are the one persistent identifier a device has that does not change.
+
+    Added in 2022.7.0.
+    """
+
+    registry = er.async_get(hass)
+    to_migrate = []
+    for entity in er.async_entries_for_config_entry(registry, entry.entry_id):
+        parts = entity.unique_id.split("_")
+        # device ID = 24 characters, MAC = 12
+        if len(parts[0]) == 24:
+            _LOGGER.debug("Entity %s needs migration", entity.entity_id)
+            to_migrate.append(entity)
+
+    if len(to_migrate) == 0:
+        _LOGGER.debug("No entities need migration to MAC address ID")
+        return
+
+    try:
+        bootstrap = await protect.get_bootstrap()
+    except (TimeoutError, ClientError, ServerDisconnectedError) as err:
+        raise ConfigEntryNotReady from err
+
+    count = 0
+    for entity in to_migrate:
+        parts = entity.unique_id.split("_")
+        device = async_device_by_id(bootstrap, parts[0])
+        if device is None:
+            continue
+
+        new_unique_id = device.mac
+        if len(parts) > 1:
+            new_unique_id = f"{device.mac}_{'_'.join(parts[1:])}"
+        _LOGGER.debug(
+            "Migrating entity %s (old unique_id: %s, new unique_id: %s)",
+            entity.entity_id,
+            entity.unique_id,
+            new_unique_id,
+        )
+        try:
+            registry.async_update_entity(entity.entity_id, new_unique_id=new_unique_id)
+        except ValueError as err:
+            print(err)
+            _LOGGER.warning(
+                "Could not migrate entity %s (old unique_id: %s, new unique_id: %s)",
+                entity.entity_id,
+                entity.unique_id,
+                new_unique_id,
+            )
+        else:
+            count += 1
+
+    if count < len(to_migrate):
+        _LOGGER.warning("Failed to migrate %s entities", len(to_migrate) - count)
```

Comment:
I had a few fail on for my Houston instance.

Reply:
Actually I think its because of the earlier buttons

Reply:
Yeah, that seems to be it.  I need to restore from back and manually clean up some of the cruft before trying again.

Reply:
Well that still blew up.  I think I'm going to have to downgrade to 2022.6.6 and then upgrade again